kernel/src/engine/default/filesystem.rs

```diff
@@ -72,7 +80,14 @@ impl<E: TaskExecutor> FileSystemClient for ObjectStoreFileSystemClient<E> {
             }
         });
 
-        Ok(Box::new(receiver.into_iter()))
+        if self.is_local {
+            // LocalFileSystem doesn't return things in the order we require
+            let mut fms: Vec<FileMeta> = receiver.into_iter().try_collect()?;
```

Comment:
What would you think of adding an optional upper bound arg to this method? 

For stores that don't guarantee order, it at least allows to filter out unwanted entries outside the requested range; for stores that _do_ guarantee order, it potentially allows pushing the bound down into the cloud store LIST call (tho I don't know of any that allow that today).

Reply:
Yep, added as an issue here #350